File: smarthomeserver/src/main/java/org/example/PersonController.java
package org.example;

import org.springframework.messaging.handler.annotation.MessageMapping;
import org.springframework.messaging.handler.annotation.SendTo;
import org.springframework.stereotype.Controller;
import org.springframework.web.util.HtmlUtils;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.atomic.AtomicInteger;

@Controller
public class PersonController {

    private Map<Integer, Player> players = new ConcurrentHashMap<>();
    private List<Integer> ids = new CopyOnWriteArrayList<>();
    private AtomicInteger counter = new AtomicInteger(0);

    @MessageMapping("/register")
    @SendTo("/topic/id")
    public Integer register(HelloMessage message) throws Exception {
        System.out.println("Request for id fromClient");
        return getNewId();
    }

    @MessageMapping("/quit")
    @SendTo("/topic/void")
    public Integer quit(Player player) throws Exception {
        System.out.println("Quit for id " + player.id);
        ids.remove(player.id);
        players.remove(player.id);
        return 0;
    }

    @MessageMapping("/updatePlayer")
    @SendTo("/topic/void")
    public Integer updatePlayer(Player player) throws Exception {
        System.out.println("Request for update id " + player.id + " fromClient");
        players.get(player.id).updatePlayer(player);
        return 0;
    }

    @MessageMapping("/getPlayers")
    @SendTo("/topic/players")
    public Collection<Player> getPlayers(HelloMessage message) throws Exception {
        System.out.println("Request for get all Players");
        return players.values();
    }

    synchronized
    private Integer getNewId(){
        Player player = new Player();
        Integer id = counter.incrementAndGet();
        player.id = id;
        ids.add(id);
        players.put(id, player);
        return id;
    }
}

File: smarthomeserver/src/main/kotlin/org/netproject/PersonController.kt
package org.netproject

import org.netproject.auth.Auth
import org.netproject.auth.SimpleAuth
import org.netproject.home.HomeState
import org.netproject.home.HomeState.Companion.update
import org.netproject.home.Roles.*
import org.netproject.home.Session
import org.netproject.message.ChangeRoleRequest
import org.netproject.message.Message
import org.netproject.message.Player
import org.springframework.messaging.handler.annotation.Headers
import org.springframework.messaging.handler.annotation.MessageMapping
import org.springframework.messaging.handler.annotation.Payload
import org.springframework.messaging.handler.annotation.SendTo
import org.springframework.messaging.simp.SimpMessagingTemplate
import org.springframework.messaging.support.MessageHeaderAccessor
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Controller
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import javax.management.relation.Role

@Controller
class PersonController(
        private val simpMessagingTemplate: SimpMessagingTemplate,
        private val homeState: HomeState,
        private val auth: Auth
) {

    @MessageMapping("/register")
    @SendTo("/topic/id")
    fun register(
            @Headers headers: MessageHeaderAccessor,
            @Payload message: Message<String>
    ): Message<String> {
        val token = "player ${UUID.randomUUID()}"
        homeState.players[headers.sessionId] = Session(token)
        return Message(message.key, token)
    }

    @MessageMapping("/changeRole")
    @SendTo("/topic/id")
    fun changeRole(
            @Headers headers: MessageHeaderAccessor,
            @Payload message: Message<ChangeRoleRequest>
    ): Message<String> = message.payload.let {
        val a = Message(message.key, auth.byPassword(headers.sessionId, it.password, Companion.byId(it.roleType)).toString())
        println(homeState.players.values.toList())
        a
    }

    @MessageMapping("/quit")
    // @SendTo("/topic/void")
    fun quit(player: Player): Int {
        val id = player.id
        // players.remove(id)
        return 0
    }

    @MessageMapping("/updatePlayer")
    fun updatePlayer(
            @Headers headers: MessageHeaderAccessor,
            @Payload message: Message<Player>) {
        homeState.players.update(headers.sessionId) { copy(player = message.payload) }
    }

    @Scheduled(fixedRate = 20)
    fun getPlayers() {
        simpMessagingTemplate.convertAndSend(
                "/topic/players",
                Message(ALL.ordinal.toString(), homeState.players.values.mapNotNull { it.player })
        )
    }

    private val MessageHeaderAccessor.sessionId: String
        get() = getHeader("simpSessionId") as String
}